File: support_tools/cpp/upd_commander_checker/src/rule_selection.cpp
#include "rule_selection.hpp"

#include <unordered_set>

namespace upd_checker {

const std::vector<std::string>& supported_rules() {
    static const std::vector<std::string> rules = {
        "UPD001", "UPD002",
        "UPD101", "UPD102", "UPD103",
        "UPD201", "UPD202", "UPD203",
        "UPD301", "UPD302", "UPD303",
        "UPD401", "UPD402", "UPD403", "UPD404"};
    return rules;
}

bool enabled_rules_are_valid(const std::vector<std::string>& rules) {
    const std::unordered_set<std::string> supported(
        supported_rules().begin(), supported_rules().end());
    for (const auto& rule : rules) {
        if (supported.find(rule) == supported.end()) {
            return false;
        }
    }
    return true;
}

std::vector<Finding> filter_enabled_findings(const RuleSelectionInput& input) {
    if (!input.enabled_rules_configured) {
        return input.findings;
    }
    const std::unordered_set<std::string> enabled(
        input.enabled_rules.begin(), input.enabled_rules.end());
    std::vector<Finding> filtered;
    for (const auto& finding : input.findings) {
        if (enabled.find(finding.code) != enabled.end()) {
            filtered.push_back(finding);
        }
    }
    return filtered;
}

}  // namespace upd_checker

File: support_tools/cs/upd_commander_checker/RuleSelection.cs
using System.Text.Json;

namespace UpdCommanderChecker;

internal static class RuleSelection
{
    internal static readonly IReadOnlyList<string> SupportedRules =
    [
        "UPD001",
        "UPD002",
        "UPD101",
        "UPD102",
        "UPD103",
        "UPD201",
        "UPD202",
        "UPD203",
        "UPD301",
        "UPD302",
        "UPD303",
        "UPD401",
        "UPD402",
        "UPD403",
        "UPD404",
        "UPD405",
        "UPD406",
    ];

    private static readonly HashSet<string> SupportedRuleSet = new(
        SupportedRules,
        StringComparer.Ordinal
    );

    internal static List<string>? ReadEnabledRules(JsonElement root)
    {
        if (!root.TryGetProperty("enabled_rules", out var property))
        {
            return null;
        }
        if (
            property.ValueKind != JsonValueKind.Array
            || property.EnumerateArray().Any(item => item.ValueKind != JsonValueKind.String)
        )
        {
            throw new ConfigException("invalid config field: enabled_rules");
        }

        var rules = property.EnumerateArray().Select(item => item.GetString()!).ToList();
        if (!AreSupported(rules))
        {
            throw new ConfigException("invalid config field: enabled_rules");
        }
        return rules;
    }

    internal static bool AreSupported(IEnumerable<string> rules)
    {
        return rules.All(SupportedRuleSet.Contains);
    }

    internal static List<Finding> Filter(RuleSelectionInput input)
    {
        if (input.EnabledRules is null)
        {
            return input.Findings.ToList();
        }

        var enabled = new HashSet<string>(input.EnabledRules, StringComparer.Ordinal);
        return input.Findings.Where(finding => enabled.Contains(finding.Code)).ToList();
    }
}
